vmm/src/memory_manager.rs

```diff
@@ -1543,22 +1554,44 @@ impl Snapshottable for MemoryManager {
         let mut memory_manager_snapshot = Snapshot::new(MEMORY_MANAGER_SNAPSHOT_ID);
         let guest_memory = self.guest_memory.memory();
 
-        let mut memory_regions: Vec<MemoryRegion> = Vec::with_capacity(10);
+        let mut memory_regions: Vec<MemoryRegion> = Vec::new();
 
         guest_memory.with_regions_mut(|index, region| {
             if region.len() == 0 {
                 return Err(MigratableError::Snapshot(anyhow!("Zero length region")));
             }
 
+            let mut backing_file = Some(PathBuf::from(format!("memory-region-{}", index)));
+            if let Some(file_offset) = region.file_offset() {
+                if (region.flags() & libc::MAP_SHARED == libc::MAP_SHARED)
+                    && Self::file_backed_with_hardlink(file_offset.file())
+                {
+                    // In this very specific case, we know the memory region
+                    // is backed by a file on the host filesystem that can be
+                    // accessed by the user, and additionally the mapping is
+                    // shared, which means that modifications to the content
+                    // are written to the actual file.
+                    // When meeting these conditions, we can skip the copy of
+                    // the memory content for this specific region, as we can
+                    // assume the user will have it saved through the backing
+                    // file already.
+                    backing_file = None;
+                }
+            }
+
             memory_regions.push(MemoryRegion {
-                backing_file: Some(PathBuf::from(format!("memory-region-{}", index))),
+                backing_file,
                 start_addr: region.start_addr(),
                 size: region.len(),
             });
 
             Ok(())
         })?;
 
+        // Store locally this list of regions as it will be used through the
+        // Transportable::send() implementation.
+        self.snapshot_memory_regions = memory_regions.clone();
```

Comment:
I understand this saves us from deserializing the Snapshot from `Transportable::send()`, so this should be more explicit in the comment. This is an optimization, not a mandatory setting.

Reply:
Ok I'll expand the comment!